hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/connector/HazelcastWriters.java

```diff
@@ -182,27 +178,36 @@ static RuntimeException handleInstanceNotActive(HazelcastInstanceNotActiveExcept
         return isLocal ? new RestartableException(e) : e;
     }
 
-    /**
-     * Wrapper class needed to conceal the JCache API while
-     * serializing/deserializing other lambdas
-     */
-    private static class CacheFlush {
+    private static class WriteCachePSupplier<K, V> extends AbstractHazelcastConnectorSupplier {
 
-        static <K, V> FunctionEx<HazelcastInstance, ConsumerEx<ArrayMap<K, V>>> flushToCache(
-                String name,
-                boolean isLocal
-        ) {
-            return instance -> {
-                ICache<K, V> cache = instance.getCacheManager().getCache(name);
-                return buffer -> {
-                    try {
-                        cache.putAll(buffer);
-                    } catch (HazelcastInstanceNotActiveException e) {
-                        throw handleInstanceNotActive(e, isLocal);
-                    }
-                    buffer.clear();
-                };
+        static final long serialVersionUID = 1L;
+
+        private final String name;
+
+        WriteCachePSupplier(@Nullable ClientConfig clientConfig, @Nonnull String name) {
+            super(asXmlString(clientConfig));
+            this.name = name;
+        }
+
+        @Override
+        protected Processor createProcessor(HazelcastInstance instance, SerializationService serializationService) {
+            FunctionEx<Context, ArrayMap<Data, Data>> bufferCreator = context -> new ArrayMap<>();
+            BiConsumerEx<ArrayMap<Data, Data>, Entry<K, V>> entryReceiver = (buffer, entry) -> {
+                Data key = serializationService.toData(entry.getKey());
+                Data value = serializationService.toData(entry.getValue());
+                buffer.add(new SimpleEntry<>(key, value));
+            };
+            ConsumerEx<ArrayMap<Data, Data>> bufferFlusher = buffer -> {
+                ICache<Data, Data> cache = instance.getCacheManager().getCache(name);
```

Comment:
It would be better to have the cache instance in the context. This is an unnecessary access to multiple concurrent maps.